File: naturalcli/src/org/naturalcli/TokenTest.java
/* 
 * TokenTest.java
 *
 * Copyright (C) 2007 Ferran Busquets
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.naturalcli;

import static org.junit.Assert.*;

import org.junit.Test;

/**
 * @author Ferran Busquets
 *
 */
public class TokenTest {

	/**
	 * Test method for {@link org.naturalcli.Token#isOptional()}.
	 */
	@Test
	public final void testSetText() throws InvalidTokenException {
		// ok
		new Token("marian");
		new Token("<marian>");
		new Token("[marian]");
		new Token("[<marian>]");
		// not ok
		try { new Token("marian]"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("marian>"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("marian>]"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("<marian"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("<marian]"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("<marian>]"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("<[marian"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("<[marian>"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("<[marian>]"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("[marian"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("[marian>"); fail(); } catch (InvalidTokenException e) { }
		try { new Token("[marian>]"); fail(); } catch (InvalidTokenException e) { }
		try { new Token(""); fail(); } catch (InvalidTokenException e) { }
		try { new Token(null); fail(); } catch (InvalidTokenException e) { }
	}
	
	/**
	 * Test method for {@link org.naturalcli.Token#isOptional()}.
	 */
	@Test
	public final void testIsOptional() throws InvalidTokenException {
		// true
		assertTrue(new Token("[marian]").isOptional());
		// false
		assertFalse(new Token("marian").isOptional());
		assertFalse(new Token("ma[]rian").isOptional());
	}

	/**
	 * Test method for {@link org.naturalcli.Token#isParameter()}.
	 */
	@Test
	public final void testIsParameter() throws InvalidTokenException {
		// true
		assertTrue(new Token("<marian>").isParameter());
		assertTrue(new Token("[<marian>]").isParameter());
		// false
		assertFalse(new Token("marian").isParameter());
		assertFalse(new Token("ma<>rian").isParameter());
	}

	/**
	 * Test method for {@link org.naturalcli.Token#isOptionalParameter()}.
	 */
	@Test
	public final void testIsOptionalParameter() throws InvalidTokenException {
		// true
		assertTrue(new Token("[<marian>]").isOptionalParameter());
		assertTrue(new Token("[<marian:integer>]").isOptionalParameter());
		// false
		assertFalse(new Token("marian").isOptionalParameter());
		assertFalse(new Token("[marian]").isOptionalParameter());
		assertFalse(new Token("<marian:integer>").isOptionalParameter());
	}
	
	/**
	 * Test method for {@link org.naturalcli.Token#getParameterName()}.
	 */
	@Test
	public final void testGetParameterName() throws InvalidTokenException {
		// null
		assertNull(new Token("marian").getParameterName());
		// equals
		assertEquals("marian", new Token("<marian>").getParameterName());
		assertEquals("marian", new Token("<marian:integer>").getParameterName());
		assertEquals("marian", new Token("<marian:email>").getParameterName());
	}

	/**
	 * Test method for {@link org.naturalcli.Token#getParameterTypeName()}.
	 */
	@Test
	public final void testGetParameterTypeName() throws InvalidTokenException {
		// null
		assertNull(new Token("marian").getParameterTypeName());
		// equals
		assertEquals("integer", new Token("<integer>").getParameterTypeName());
		assertEquals("integer", new Token("<marian:integer>").getParameterTypeName());
		assertEquals("email", new Token("<marian:email>").getParameterTypeName());
	}

	/**
	 * Test method for {@link org.naturalcli.Token#isIdentifier()}.
	 */
	@Test
	public final void testIsIdentifier() throws InvalidTokenException {
		// Ok
		assertTrue(new Token("marian").isIdentifier());
		assertTrue(new Token("[marian]").isIdentifier());
		// Not ok
		assertFalse(new Token("<marian>").isIdentifier());
		assertFalse(new Token("[<marian>]").isIdentifier());
	}

	/**
	 * Test method for {@link org.naturalcli.Token#getWord()}.
	 */
	@Test
	public final void testGetWord() throws InvalidTokenException {
		assertEquals("marian", new Token("<marian>").getWord());
		assertEquals("marian", new Token("[marian]").getWord());
		assertEquals("marian:integer", new Token("<marian:integer>").getWord());
		assertEquals("marian:email", new Token("[<marian:email>]").getWord());
	}

	/**
	 * Test method for {@link org.naturalcli.Token#matches(java.lang.String, org.naturalcli.parameters.ParameterValidator)}.
	 * @throws UnknownParameterType 
	 */
	@Test
	public final void testMatches() throws InvalidTokenException, UnknownParameterType {
		ParameterValidator pv = new ParameterValidator();
		// Ok
		assertTrue(new Token("marian").matches("marian", pv));
		assertTrue(new Token("<integer>").matches("1234", pv));
		assertTrue(new Token("<marian:email>").matches("marian@marian.org", pv));
		// Not ok
		assertFalse(new Token("marian").matches("1234", pv));
		assertFalse(new Token("marian").matches(null, pv));
		assertFalse(new Token("marian").matches("", pv));
		assertFalse(new Token("<integer>").matches("marian", pv));
		assertFalse(new Token("<marian:email>").matches("marian", pv));
	}

}

File: naturalcli_python/test/naturalcli/commands/TokenTest.py
import unittest
import naturalcli


class TokenTest (unittest.TestCase):  

    """
      Test method for {@link org.naturalcli.Token#isOptional()}.
    """
    
    def testSetText(self):
        # ok
        Token("marian")
        Token("<marian>")
        Token("[marian]")
        Token("[<marian>]")
        # not ok
        try :
            Token("marian]")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("marian>")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("marian>]") 
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("<marian")
            fail() 
        except InvalidTokenError:
            pass
        
        try :
            Token("<marian]") 
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("<marian>]")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("<[marian")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("<[marian>")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("<[marian>]")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("[marian")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("[marian>")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("[marian>]")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token("")
            fail() 
        except InvalidTokenError:
            pass
        
        try : 
            Token(None)
            fail() 
        except InvalidTokenError:
            pass

    
    def testIsOptional(self):
        """
          Test method for {@link org.naturalcli.Token#isOptional()}.
        """
        # true
        assert Token("[marian]").isOptional()
        # false
        assert Token("marian").isOptional()
        assert Token("ma[]rian").isOptional()
    
    def testIsParameter(self):
        """
          Test method for {@link org.naturalcli.Token#isParameter()}.
        """
        # true
        assert Token("<marian>").isParameter()
        assert Token("[<marian>]").isParameter()
        # false
        assert Token("marian").isParameter()
        assert Token("ma<>rian").isParameter()

    
    def testIsOptionalParameter(self):
        """
          Test method for {@link org.naturalcli.Token#isOptionalParameter()}.
        """
        # true
        assert Token("[<marian>]").isOptionalParameter()
        assert Token("[<marian:integer>]").isOptionalParameter()
        # false
        assert Token("marian").isOptionalParameter()
        assert Token("[marian]").isOptionalParameter()
        assert Token("<marian:integer>").isOptionalParameter()

    """
      Test method for {@link org.naturalcli.Token#getParameterName()}.
    """
    
    def testGetParameterName(self):
        # None
        assert None == Token("marian").getParameterName()
        # equals
        assert "marian" == Token("<marian>").getParameterName()
        assert "marian" == Token("<marian:integer>").getParameterName()
        assert "marian" == Token("<marian:email>").getParameterName()

    
    def testGetParameterTypeName(self):
        """
          Test method for {@link org.naturalcli.Token#getParameterTypeName()}.
        """
        # None
        assert None == Token("marian").getParameterTypeName()
        # equals
        assert "integer" == Token("<integer>").getParameterTypeName()
        assert "integer" == Token("<marian:integer>").getParameterTypeName()
        assert "email" == Token("<marian:email>").getParameterTypeName()

    
    def testIsIdentifier(self):
        """
          Test method for {@link org.naturalcli.Token#isIdentifier()}.
        """
        # Ok
        assert Token("marian").isIdentifier()
        assert Token("[marian]").isIdentifier()
        # Not ok
        assert not Token("<marian>").isIdentifier()
        assert not Token("[<marian>]").isIdentifier()

    
    def testGetWord(self):
        """
          Test method for {@link org.naturalcli.Token#getWord()}.
        """
        assert "marian" == Token("<marian>").getWord())
        assert "marian" == Token("[marian]").getWord())
        assert "marian:integer" == Token("<marian:integer>").getWord())
        assert "marian:email" == Token("[<marian:email>]").getWord())

    """
      Test method for {@link org.naturalcli.Token#matches(java.lang.String, org.naturalcli.parameters.ParameterValidator)}.
      @throws UnknownParameterType 
    """
    
    def testMatches(self):
        ParameterValidator pv = new ParameterValidator()
        # Ok
        assert Token("marian").matches("marian", pv))
        assert Token("<integer>").matches("1234", pv))
        assert Token("<marian:email>").matches("marian@marian.org", pv))
        # Not ok
        assert not Token("marian").matches("1234", pv))
        assert not Token("marian").matches(None, pv))
        assert not Token("marian").matches("", pv))
        assert not Token("<integer>").matches("marian", pv))
        assert not Token("<marian:email>").matches("marian", pv))
